src/util/Timer.cpp

```diff
@@ -16,26 +17,46 @@ namespace stellar
 using namespace std;
 
 static const uint32_t RECENT_CRANK_WINDOW = 1024;
+static const std::chrono::milliseconds CRANK_TIME_SLICE(500);
+static const size_t CRANK_EVENT_SLICE = 100;
+static const std::chrono::seconds SCHEDULER_LATENCY_WINDOW(5);
 
-VirtualClock::VirtualClock(Mode mode) : mMode(mode), mRealTimer(mIOContext)
+VirtualClock::VirtualClock(Mode mode)
+    : mMode(mode)
+    , mActionScheduler(
+          std::make_unique<Scheduler>(*this, SCHEDULER_LATENCY_WINDOW))
+    , mRealTimer(mIOContext)
 {
-    mExecutionIterator = mExecutionQueue.begin();
     resetIdleCrankPercent();
 }
 
 VirtualClock::time_point
-VirtualClock::now() noexcept
+VirtualClock::now() const noexcept
 {
     if (mMode == REAL_TIME)
     {
-        return std::chrono::system_clock::now();
+        return std::chrono::steady_clock::now();
     }
     else
     {
         return mVirtualNow;
     }
 }
 
+VirtualClock::system_time_point
+VirtualClock::system_now() const noexcept
+{
+    if (mMode == REAL_TIME)
+    {
+        return std::chrono::system_clock::now();
+    }
+    else
+    {
+        auto offset = mVirtualNow.time_since_epoch();
+        return std::chrono::system_clock::time_point(offset);
```

Comment:
I don't think there's any overflow risk here; if I'm reading correctly:
~~~
std::nano == std::ratio<1, 1000000000>`
system_clock::duration::period
      == std::ratio_multiply<std::ratio<100,1>, std::nano>
      == std::ratio<1, 10000000>
~~~
That is, they're defining a `system_clock::duration` with `period` = 100ns rather than 1ns; and doing a `duration_cast` from `steady_clock` with 1ns period to `system_clock` with 100ns period is just going to cause a division-by-100.